Add predefined ImportOption to exclude package-info.class files

TomerFi opened the pull request:
Resolves #793

hankem commented on archunit/src/main/java/com/tngtech/archunit/core/importer/ImportOption.java:
```diff
@@ -137,4 +149,16 @@ public boolean includes(Location location) {
+    /**
+     * The {@code maven-compiler-plugin 3.10.0} fixed a bug that made {@code package-info.class}
+     * files hidden from {@code ArchUnit}.
+     * This is a {@link ImportOption} impelmentation for excluding {@code package-info.class} files.
```
There's a typo in `impelmentation`.

codecholeric commented on archunit/src/main/java/com/tngtech/archunit/core/importer/ImportOption.java:
```diff
@@ -137,4 +149,16 @@ public boolean includes(Location location) {
     }
+
+    /**
+     * The {@code maven-compiler-plugin 3.10.0} fixed a bug that made {@code package-info.class}
```
I don't think we need this context information inside of the Javadoc? This sounds to me like the reason you stumbled over it, but `package-info.class` has been there long before, I have already used it with ArchUnit, Gradle and IntelliJ long ago :man_shrugging: This is for me context information that could go into the commit message or PR message, but I don't think it's relevant for users who want to read the Javadoc of this `ImportOption`. In the Javadoc I would rather expect to see an explanation what `package-info` actually is (maybe with a link to the JLS) and that this `ImportOption` filters it out :thinking:

TomerFi replied:
good point.

TomerFi replied:
good point.
cleaned it up in https://github.com/TNG/ArchUnit/pull/794/commits/90c126f6acba68e0602e316c5d576001ee5f20ce

codecholeric commented on archunit/src/main/java/com/tngtech/archunit/core/importer/ImportOption.java:
```diff
@@ -137,4 +149,16 @@ public boolean includes(Location location) {
+    final class DoNotIncludePackageInfoClasses implements ImportOption {
+        @Override
+        public boolean includes(Location location) {
+            return !location.matches(Pattern.compile(".*package-info\\.class$"));
```
We should not call `Pattern.compile(..)` on every `includes(..)` invocation. That is quite inefficient. We should compile the `Pattern` one time into a static constant and then reuse it on every `includes(..)` invocation (which might in the end be called tens or hundreds of thousands of times for big projects).

I was also wondering if we should add a simple `Location.endsWith(..)` public API to save some performance :thinking: But I have realized that we're (ab-)using patterns also for some other simple cases like `MAVEN_TEST_PATTERN`. Don't know if there is an internal optimization that makes this as efficient as `contains`, maybe we should do some measurements at some point (unrelated to this PR :wink:)

TomerFi replied:
yeah, I guess I missed this, good catch.
fixed in https://github.com/TNG/ArchUnit/pull/794/commits/90c126f6acba68e0602e316c5d576001ee5f20ce

codecholeric commented on archunit/src/test/java/com/tngtech/archunit/core/importer/ImportOptionsTest.java:
```diff
@@ -157,6 +161,30 @@ public void detects_archives_correctly(ImportOption doNotIncludeArchives) {
+    @Test
+    @UseDataProvider("do_not_include_package_info_classes")
+    public void detect_package_info_class(ImportOption doNotIncludePackageInfoClasses) {
+        NormalizedUri mockUri = mock(NormalizedUri.class);
```
I would not mock things like this. I think mocking is great for collaborators, but not great for domain objects (you lose intrinsic invariants, i.e. you can create unrealistic objects, etc., and the tests are harder to maintain on refactorings).
Is there any reason you don't simply use something like

```
Location location = Location.of(URI.create("file:///path/to/package/and/file/package-info.class"));
```

Then the whole extra class `StubLocation` is also obsolete... :man_shrugging:

TomerFi replied:
that's funny - at the start, I actually did something similar to this and I changed it to the stubbing class instead of including an extra file.
but you are right - it is better like that.
fixed in https://github.com/TNG/ArchUnit/pull/794/commits/90c126f6acba68e0602e316c5d576001ee5f20ce

codecholeric commented on archunit/src/test/java/com/tngtech/archunit/core/importer/ImportOptionsTest.java:
```diff
@@ -157,6 +161,30 @@ public void detects_archives_correctly(ImportOption doNotIncludeArchives) {
+        given(mockUri.toString())
+                .willReturn("/path/to/package/and/file/package-info.class");
+        assertThat(doNotIncludePackageInfoClasses.includes(location))
+                .as("with path for package-info.class file")
```
I think it's better if the description really refers to the argument of `assertThat(..)`. Because this way you would get an error like `with path for package-info.class file > expected false but was true`. But what should that mean? It would likely be better to have sth. like `doNotIncludePackageInfoClasses.includes(locationOfPackageInfo)`, then seeing `expected false but was true` makes sense, no? :thinking:

TomerFi replied:
yes, makes perfect sense.
fixed in https://github.com/TNG/ArchUnit/pull/794/commits/90c126f6acba68e0602e316c5d576001ee5f20ce

codecholeric commented on archunit/src/main/java/com/tngtech/archunit/core/importer/ImportOption.java:
```diff
@@ -137,4 +149,16 @@ public boolean includes(Location location) {
+     * files hidden from {@code ArchUnit}.
+     * This is a {@link ImportOption} implementation for excluding {@code package-info.class} files.
+     */
+    final class DoNotIncludePackageInfoClasses implements ImportOption {
```
Do you think we really need the suffix `Classes`? :thinking: Or would `DoNotIncludePackageInfos` be good enough? Just wondering what is consistent, because it's also called `DoNotIncludeTests` and not `DoNotIncludeTestClasses` :thinking:

TomerFi replied:
good point.
fixed in https://github.com/TNG/ArchUnit/pull/794/commits/90c126f6acba68e0602e316c5d576001ee5f20ce

codecholeric commented on archunit/src/test/java/com/tngtech/archunit/core/importer/ImportOptionsTest.java:
```diff
@@ -186,4 +214,31 @@ FolderPattern expectTestFolder() {
         }
     }
+
+    private static class StubLocation extends Location {
```
As mentioned I think we can do without this :slightly_smiling_face:

TomerFi replied:
this class is no longer with us.
fixed in https://github.com/TNG/ArchUnit/pull/794/commits/90c126f6acba68e0602e316c5d576001ee5f20ce

codecholeric commented on archunit/src/main/java/com/tngtech/archunit/core/importer/ImportOption.java:
```diff
@@ -76,6 +76,18 @@ public boolean includes(Location location) {
+        },
+
+        /**
+         * Do not include package-info.class files
```
Reading the Javadocs of these classes now I think the best thing would be to write comprehensive Javadoc on the class and then simply `@see ImportOptionClass` here :thinking: Because `Predefined` and the classes themselves (that you would use with JUnit testing support) should give the same information. Unfortunately the given Javadocs are not exactly great examples :joy: We should clean that up at some point and always refer from `Predefined` to the class that then has the full Javadoc...

codecholeric reviewed (changes requested):
Thanks a lot for your contribution :smile: As always I have way to many review comments :see_no_evil: I hope you're still motivated to look into them and let me know what you think!

TomerFi commented:
> Thanks a lot for your contribution smile As always I have way to many review comments see_no_evil I hope you're still motivated to look into them and let me know what you think!

Not a problem - a good code review is always a good thing.
I'll make the required adjustments by the weekend, most likely tomorrow.

codecholeric reviewed (approved):
Cool, thanks a lot :smile: I rebased on `main` and fixed it up a tiny bit to be consistent with existing style (e.g. we don't use prefixes for the commit message and it shouldn't exceed 70 chars). Hope that's okay with you! (just wanna make sure it ends up in the next release :slightly_smiling_face:)

TomerFi commented:
> Cool, thanks a lot smile I rebased on `main` and fixed it up a tiny bit to be consistent with existing style (e.g. we don't use prefixes for the commit message and it shouldn't exceed 70 chars). Hope that's okay with you! (just wanna make sure it ends up in the next release slightly_smiling_face)

@codecholeric  Sure, it's your repo - thanks a lot for sharing it and accepting my small contribution.
😻 

I'm just curious...

Because for my projects, I usually do the exact opposite.
I'm a big fan of [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/),
I use the prefix (also known as a type) to categorize my commits, it helps me automate version bumps.
I also used to automate changelog in the same manner,
but today GH releases do an awesome changelog based on PRs.

May I ask why the 70 chars limit?
Does it work well with dependabo'st PRs?
